File: src/engine.rs
use raylib::prelude::*;
use std::{
    ops::Deref,
    sync::{Arc, Mutex, MutexGuard},
};

/// # paper-engine
// asdf

type GridCell = Vec<Option<Arc<Mutex<Particle>>>>;

pub static GRAVITY: f32 = 9.81;

pub fn get_sign(number: f32) -> f32 {
    if number > 0.0 {
        1.0
    } else if number < 0.0 {
        -1.0
    } else {
        0.0
    }
}

#[derive(Debug)]
pub struct Grid {
    /// width and height are the width and height of each cell
    pub width: i32,
    pub height: i32,
    // Arc of two-dimensional arraylist that contains cells
    pub content: Arc<Mutex<Vec<Vec<GridCell>>>>,
}

impl Grid {
    pub fn new() -> Self {
        let width = 20;
        let height = 20;

        Self {
            width,
            height,
            content: Arc::new(Mutex::new(vec![])),
        }
    }

    /// Finds collisions with adjacent cells
    pub fn find_collisions(&mut self) {
        // x and y start at one to skip topmost and bottommost rows, and leftmost and rightmost columns
        for x in 1..=self.width {
            for y in 1..=self.height {
                for dx in -1..=1 {
                    for dy in -1..=1 {
                        self.solve_collision((x, y), (x + dx, y + dy));
                    }
                }
            }
        }
    }

    fn solve_collision(
        &mut self,
        (cell_a_x, cell_a_y): (i32, i32),
        (cell_b_x, cell_b_y): (i32, i32),
    ) {
        let cell_a = &mut self.content.lock().unwrap()[cell_a_x as usize][cell_a_y as usize];
        let cell_b = &mut self.content.lock().unwrap()[cell_b_x as usize][cell_b_y as usize];

        for particle_a_wrapped in cell_a.iter_mut() {
            for particle_b_wrapped in cell_b.iter_mut() {
                if !std::ptr::eq(particle_a_wrapped, particle_b_wrapped) {
                    if let Some(a_refcell) = particle_a_wrapped.to_owned() {
                        if let Some(b_refcell) = particle_b_wrapped.to_owned() {
                            let mut a = a_refcell.lock().unwrap();
                            let mut b = b_refcell.lock().unwrap();

                            if check_collision_circles(a.pos, a.radius, b.pos, b.radius) {
                                let compensation_length =
                                    a.radius + b.radius - a.pos.distance_to(b.pos);
                                a.pos -= compensation_length / 2.0;
                                b.pos += compensation_length / 2.0;
                            }
                        }
                    }
                }
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct Particle {
    pub pos: Vector2,
    pub vel: Vector2,
    pub radius: f32,
    collision_damping: f32,
}

impl Particle {
    pub fn new(start_pos: Vector2, vel: Vector2, radius: f32) -> Self {
        Self {
            pos: start_pos,
            vel,
            radius,
            collision_damping: 0.8,
        }
    }

    pub fn collision(&mut self, bounds: &Rectangle) {
        let bounds_size = Vector2::new(bounds.x + bounds.width, bounds.y + bounds.height)
            - Vector2::one() * self.radius;

        if self.pos.x > bounds_size.x {
            self.pos.x = bounds_size.x * get_sign(self.pos.x);
            self.vel.x *= -1.0 * self.collision_damping;
        }

        if self.pos.y > bounds_size.y {
            self.pos.y = bounds_size.y * get_sign(self.pos.y);
            self.vel.y *= -1.0 * self.collision_damping;
        }
    }

    fn in_cell(&self, grid: &Grid, cell: (i32, i32)) -> bool {
        if self.pos.x > (grid.width * cell.0) as f32
            && self.pos.x < (grid.width * cell.0 + grid.width) as f32
            && self.pos.y > (grid.width * cell.1 + grid.height) as f32
            && self.pos.y < (grid.width * cell.1) as f32
        {
            true
        } else {
            false
        }
    }

    pub fn update(&mut self, dh: &mut RaylibDrawHandle, bounds: &Rectangle, grid: &mut Grid) {
        self.vel.y += GRAVITY;

        self.pos += self.vel * dh.get_frame_time();
        self.collision(&bounds);

        let g = grid.content.to_owned();

        for (i, row) in g.lock().unwrap().iter_mut().enumerate() {
            for (j, c) in row.iter_mut().enumerate() {
                let mut indices: Vec<usize> = vec![];
                let cell = Arc::new(Mutex::new(c));
                for (k, p) in cell.lock().unwrap().iter().enumerate() {
                    if let Some(arc_mutex) = p {
                        if let Ok(mut particle) = arc_mutex.lock() {
                            let particle_ref = &mut *particle;
                            let particle_addr = particle_ref as *mut Particle;
                            let self_addr = self as *mut Particle;

                            if particle_addr == self_addr {
                                indices.push(k);
                            }
                        }
                    }

                    // Mutate self in grid
                    for &k in indices.iter().rev() {
                        g.lock().unwrap()[(self.pos.y / grid.height as f32) as usize]
                            [(self.pos.x / grid.width as f32) as usize]
                            .push(cell.lock().unwrap().swap_remove(k));
                    }
                }
            }

            dh.draw_circle(
                self.pos.x as i32,
                self.pos.y as i32,
                self.radius,
                Color::LIGHTBLUE,
            );
        }
    }
}

File: engine.go
package main

import (
	"fmt"

	"github.com/gen2brain/raylib-go/raylib"
)

type Vector2 = rl.Vector2
type Vector3 = rl.Vector3

const GRAVITY float32 = 9.81

func Vector3MultiplyValue(vec Vector3, val float32) Vector3 {
	return rl.NewVector3(vec.X*val, vec.Y*val, vec.Z*val)
}

func Vector3to2(vec rl.Vector3) rl.Vector2 {
	return rl.NewVector2(vec.X, vec.Y)
}

func GetSign(val float32) float32 {
	if val > 0 {
		return 1
	} else if val < 0 {
		return -1
	} else {
		return 0
	}
}

type Particle struct {
	Pos              rl.Vector3
	Vel              rl.Vector3
	Radius           float32
	CollisionDamping float32
	Color            rl.Color
	Id               int
}

func NewParticle(pos, vel Vector3, radius float32, color rl.Color, id int) *Particle {
	return &Particle{
		Pos:              pos,
		Vel:              vel,
		Radius:           radius,
		CollisionDamping: 0.8,
		Color:            color,
		Id:               id,
	}
}

func (self *Particle) Update(bounds *Bounds) {
	self.Vel.Y -= GRAVITY * rl.GetFrameTime()
	self.Pos = rl.Vector3Add(self.Pos, self.Vel)

	if self.Pos.Y < bounds.YMin {
		self.Pos.Y = bounds.Pos.Y + (bounds.Height/2)*GetSign(self.Pos.Y)
		self.Vel.Y *= -1 * self.CollisionDamping
	} else if self.Pos.Y > bounds.YMax {
		self.Pos.Y = bounds.Pos.Y - (bounds.Height/2)*GetSign(self.Pos.Y)
		self.Vel.Y *= -1 * self.CollisionDamping
	}

	if self.Pos.X < bounds.XMin {
		self.Pos.X = bounds.Pos.X + (bounds.Width/2)*GetSign(self.Pos.X)
		self.Vel.X *= -1 * self.CollisionDamping
	} else if self.Pos.X > bounds.XMax {
		self.Pos.X = bounds.Pos.X - (bounds.Width/2)*GetSign(self.Pos.X)
		self.Vel.X *= -1 * self.CollisionDamping
	}

	if self.Pos.Z < bounds.ZMin {
		self.Pos.Z = bounds.Pos.Z + (bounds.Length/2)*GetSign(self.Pos.Z)
		self.Vel.Z *= -1 * self.CollisionDamping
	} else if self.Pos.Z > bounds.ZMax {
		self.Pos.Z = bounds.Pos.Z - (bounds.Length/2)*GetSign(self.Pos.Z)
		self.Vel.Z *= -1 * self.CollisionDamping
	}

	rl.DrawSphere(self.Pos, self.Radius, self.Color)
}

func (self *Particle) Collision() {}

type Bounds struct {
	Pos    Vector3
	Width  float32
	Height float32
	Length float32

	YMin float32
	YMax float32
	XMin float32
	XMax float32
	ZMin float32
	ZMax float32
}

func NewBounds(pos Vector3, width, height, length float32) Bounds {
	fmt.Print()
	return Bounds{
		Pos:    pos,
		Width:  width,
		Height: height,
		Length: length,

		YMin: pos.Y - height/2,
		YMax: pos.Y + height/2,
		XMin: pos.X - width/2,
		XMax: pos.X + width/2,
		ZMin: pos.Z - length/2,
		ZMax: pos.Z + length/2,
	}
}

type Grid struct {
	Width  float32
	Height float32
	Length float32
}
